pkg/autoscaler/scaling/autoscaler.go

```diff
@@ -182,45 +184,42 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (desiredPodCount
 		dspc, dppc, originalReadyPodsCount, maxScaleUp, maxScaleDown)
 
 	// We want to keep desired pod count in the  [maxScaleDown, maxScaleUp] range.
-	desiredStablePodCount := int32(math.Min(math.Max(dspc, maxScaleDown), maxScaleUp))
-	desiredPanicPodCount := int32(math.Min(math.Max(dppc, maxScaleDown), maxScaleUp))
+	desiredStablePodCount := int32(clampInclusiveRange(dspc, maxScaleDown, maxScaleUp))
+	desiredPanicPodCount := int32(clampInclusiveRange(dppc, maxScaleDown, maxScaleUp))
 
 	logger.With(zap.String("mode", "stable")).Debugf("Observed average scaling metric value: %0.3f, targeting %0.3f.",
 		observedStableValue, spec.TargetValue)
 	logger.With(zap.String("mode", "panic")).Debugf("Observed average scaling metric value: %0.3f, targeting %0.3f.",
 		observedPanicValue, spec.TargetValue)
 
 	isOverPanicThreshold := observedPanicValue/readyPodsCount >= spec.PanicThreshold
+	isBelowPanicThreshold := !isOverPanicThreshold
 
 	a.stateMux.Lock()
 	defer a.stateMux.Unlock()
-	if a.panicTime.IsZero() && isOverPanicThreshold {
-		// Begin panicking when we cross the threshold in the panic window.
-		logger.Info("PANICKING.")
-		a.panicTime = now
-		pkgmetrics.Record(a.reporterCtx, panicM.M(1))
-	} else if !a.panicTime.IsZero() && !isOverPanicThreshold && a.panicTime.Add(spec.StableWindow).Before(now) {
-		// Stop panicking after the surge has made its way into the stable metric.
-		logger.Info("Un-panicking.")
-		a.panicTime = time.Time{}
-		a.maxPanicPods = 0
-		pkgmetrics.Record(a.reporterCtx, panicM.M(0))
+	if isOverPanicThreshold && a.currentlyStable() {
+		a.startPanicking(now, logger)
+	} else if isBelowPanicThreshold && a.currentlyPanicking() && a.hasElapsed(now, spec.StableWindow) {
+		a.stopPanicking(logger)
 	}
 
-	if !a.panicTime.IsZero() {
+	if a.currentlyStable() {
+		logger.Debug("Operating in stable mode.")
+
+		desiredPodCount = desiredStablePodCount
+	} else if a.currentlyPanicking() {
```

Comment:
The `else if` here seems redundant and even makes me ask if there are more states than "stable" and "panicking". I'd leave this just be `else`.

Reply:
Yes, but the naked `else` assumes that I already know what the other state is. This makes sense when there is only true/false, but actually this is mode/other-mode.

Reply:
That might be one of the reasons to inline the check? :)

Reply:
The logic here isn't really about whether a number is zero, it's about two distinct modes of operation. That's why I extracted the query methods, to put a name on that distinction.